swap/src/protocol/bob.rs

```diff
@@ -3,30 +3,31 @@ use std::sync::Arc;
 use anyhow::Result;
 use uuid::Uuid;
 
-use crate::database::Database;
+use crate::protocol::Database;
 use crate::{bitcoin, cli, env, monero};
 
 pub use self::state::*;
 pub use self::swap::{run, run_until};
+use std::convert::TryInto;
 
 pub mod state;
 pub mod swap;
 
-pub struct Swap {
+pub struct Swap<D: Database> {
```

Comment:
Same here, unless you are accessing associated types of a type parameter, try to avoid putting bounds on them in the struct declaration.